// @react-awesome-query-builder/ui/cjs/components/rule/WidgetFactory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _core = require("@react-awesome-query-builder/core");
var _omit = _interopRequireDefault(require("lodash/omit"));
var _excluded = ["factory"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getTitleInListValues = _core.Utils.ListUtils.getTitleInListValues;
var _widgetDefKeysToOmit = _core.Utils.ConfigUtils._widgetDefKeysToOmit;
var _fixImmutableValue = _core.Utils.TreeUtils._fixImmutableValue;
var _default = exports["default"] = function _default(_ref) {
  var delta = _ref.delta,
    isFuncArg = _ref.isFuncArg,
    valueSrc = _ref.valueSrc,
    immValue = _ref.value,
    immValueError = _ref.valueError,
    fieldError = _ref.fieldError,
    asyncListValues = _ref.asyncListValues,
    isSpecialRange = _ref.isSpecialRange,
    fieldDefinition = _ref.fieldDefinition,
    widget = _ref.widget,
    widgetDefinition = _ref.widgetDefinition,
    widgetValueLabel = _ref.widgetValueLabel,
    valueLabels = _ref.valueLabels,
    textSeparators = _ref.textSeparators,
    setValueHandler = _ref.setValueHandler,
    setFuncValue = _ref.setFuncValue,
    config = _ref.config,
    field = _ref.field,
    fieldSrc = _ref.fieldSrc,
    fieldType = _ref.fieldType,
    isLHS = _ref.isLHS,
    operator = _ref.operator,
    readonly = _ref.readonly,
    parentField = _ref.parentField,
    parentFuncs = _ref.parentFuncs,
    id = _ref.id,
    groupId = _ref.groupId,
    widgetId = _ref.widgetId;
  var widgetFactory = widgetDefinition.factory,
    fieldWidgetProps = (0, _objectWithoutProperties2["default"])(widgetDefinition, _excluded);
  var isConst = isFuncArg && fieldDefinition.valueSources && fieldDefinition.valueSources.length == 1 && fieldDefinition.valueSources[0] == "const";
  var defaultValue = fieldDefinition.defaultValue;
  if (!widgetFactory) {
    return "?";
  }

  // Widget value (if it's not a func with args) should NOT be Immutable
  // Eg. for multiselect value should be a JS Array, not Immutable List
  var fixedImmValue = immValue ? immValue.map(function (v) {
    return _fixImmutableValue(v);
  }) : undefined;
  var value = isSpecialRange ? [fixedImmValue === null || fixedImmValue === void 0 ? void 0 : fixedImmValue.get(0), fixedImmValue === null || fixedImmValue === void 0 ? void 0 : fixedImmValue.get(1)] : fixedImmValue ? fixedImmValue.get(delta) : undefined;
  var valueError = (immValueError === null || immValueError === void 0 ? void 0 : immValueError.get) && (isSpecialRange ? [immValueError.get(0), immValueError.get(1), immValueError.get(2)] : immValueError.get(delta)) || null;
  var errorMessage = isLHS ? fieldError : valueError;
  if (isSpecialRange && value[0] === undefined && value[1] === undefined) value = undefined;
  var _ref2 = fieldDefinition || {},
    fieldSettings = _ref2.fieldSettings;
  var widgetProps = (0, _omit["default"])(_objectSpread(_objectSpread(_objectSpread({}, fieldWidgetProps), fieldSettings), {}, {
    config: config,
    field: field,
    fieldSrc: fieldSrc,
    fieldType: fieldType,
    isLHS: isLHS,
    parentField: parentField,
    parentFuncs: parentFuncs,
    fieldDefinition: fieldDefinition,
    operator: operator,
    delta: delta,
    isSpecialRange: isSpecialRange,
    isFuncArg: isFuncArg,
    value: value,
    valueError: valueError,
    fieldError: fieldError,
    errorMessage: errorMessage,
    label: widgetValueLabel.label,
    placeholder: widgetValueLabel.placeholder,
    placeholders: valueLabels ? valueLabels.placeholder : null,
    textSeparators: textSeparators,
    setValue: setValueHandler,
    setFuncValue: setFuncValue,
    readonly: readonly,
    asyncListValues: asyncListValues,
    id: id,
    groupId: groupId,
    widgetId: widgetId
  }), [].concat((0, _toConsumableArray2["default"])(_widgetDefKeysToOmit), ["toJS"]));
  if (widget == "field") {
    //
  }
  if (isConst && defaultValue) {
    var listValues = (fieldSettings === null || fieldSettings === void 0 ? void 0 : fieldSettings.treeValues) || (fieldSettings === null || fieldSettings === void 0 ? void 0 : fieldSettings.listValues);
    if (typeof defaultValue == "boolean") {
      return defaultValue ? widgetProps.labelYes || "YES" : widgetProps.labelNo || "NO";
    } else if (listValues) {
      if (Array.isArray(defaultValue)) return defaultValue.map(function (v) {
        return getTitleInListValues(listValues, v) || v;
      }).join(", ");else return getTitleInListValues(listValues, defaultValue) || defaultValue;
    }
    return "" + defaultValue;
  }
  return widgetFactory(widgetProps, config.ctx);
};
